Class/practice6.1/MessageDispatcher.h

```diff
@@ -0,0 +1,55 @@
+#ifndef MESSAGEDISPATCHER_H
+#define MESSAGEDISPATCHER_H
+
+#include <iostream>
+#include <functional>
+#include <vector>
+#include <mutex>
+#include "Messages.h"
+
+class MessageDispatcher {
+public:
+    void publish(const GreenMessage& message) {
+        std::lock_guard<std::mutex> lock(mutex_);
+        for (const auto& subscriber : greenSubscribers_) {
+            subscriber(message);
+        }
+    }
+
+    void publish(const BlueMessage& message) {
+        std::lock_guard<std::mutex> lock(mutex_);
+        for (const auto& subscriber : blueSubscribers_) {
+            subscriber(message);
```

Comment:
There is a problem here. publish method is expected to be (almost)instant while in your implementation it should be called publishAndHandle not publish. The main issue here is that we don't know what callbacks will be registered and how long it will take to wait for their completion. Therefore publisher thread will be busy to do all that job instead of just publish message to queue and continue work(it is critical if we have GUI events or something like that).
The regular approach is to have queue or queues to store the messages and dedicated worker threads to feed the messages into subscribers